gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/ManifestBasedDataset.java

```diff
@@ -0,0 +1,145 @@
+  private static final String DELETE_FILE_NOT_EXIST_ON_SOURCE = CONFIG_PREFIX + ".deleteFileNotExistOnSource";
+  /** If true, will delete newly empty directories up to the config set in DELETE_EMPTY_DIRECTORIES_UPTO. */
+  public static final String DELETE_EMPTY_DIRECTORIES_KEY = CONFIG_PREFIX + ".deleteEmptyDirectories";
+  public static final String DELETE_EMPTY_DIRECTORIES_UPTO = CONFIG_PREFIX + ".deleteEmptyDirectoriesUpTo";
+  private final FileSystem fs;
+  private final Path manifestPath;
+  private final Properties properties;
+  private final boolean deleteEmptyDirectories;
+  private Gson GSON = new Gson();
+
+  public ManifestBasedDataset(final FileSystem fs, Path manifestPath, Properties properties) {
+    this.fs = fs;
+    this.manifestPath = manifestPath;
+    this.properties = properties;
+    this.deleteEmptyDirectories = Boolean.parseBoolean(properties.getProperty(DELETE_EMPTY_DIRECTORIES_KEY, "false"))
+        && properties.containsKey(DELETE_EMPTY_DIRECTORIES_UPTO);
+  }
+
+  @Override
+  public String datasetURN() {
+    return this.manifestPath.toString();
+  }
+
+  @Override
+  public Iterator<FileSet<CopyEntity>> getFileSetIterator(FileSystem targetFs, CopyConfiguration configuration)
+      throws IOException {
+    if (!fs.exists(manifestPath)) {
+      log.warn(String.format("Manifest path %s does not exist on filesystem %s, will not copy data in this manifest"
+          + ", probably due to wrong configuration of %s", manifestPath.toString(), fs.getUri().toString(), ManifestBasedDatasetFinder.MANIFEST_LOCATION));
+      return Iterators.emptyIterator();
+    } else if (fs.getFileStatus(manifestPath).isDirectory()) {
+      log.warn(String.format("Manifest path %s on filesystem %s is a directory, which is not supported. Please set the manifest file locations in"
+          + "%s, you can specify multi locations split by '',", manifestPath.toString(), fs.getUri().toString(), ManifestBasedDatasetFinder.MANIFEST_LOCATION));
+      return Iterators.emptyIterator();
+    }
+    JsonReader reader = null;
+    List<CopyEntity> copyEntities = Lists.newArrayList();
+    List<FileStatus> toDelete = Lists.newArrayList();
+    //todo: put permission preserve logic here?
+    try {
+      reader = new JsonReader(new InputStreamReader(fs.open(manifestPath), "UTF-8"));
+      reader.beginArray();
+      while (reader.hasNext()) {
+        //todo: We can use fileSet to partition the data in case of some softbound issue
+        //todo: After partition, change this to directly return iterator so that we can save time if we meet resources limitation
+        JsonObject file = GSON.fromJson(reader, JsonObject.class);
+        Path fileToCopy = new Path(file.get("fileName").getAsString());
+        if (this.fs.exists(fileToCopy)) {
+          if (!targetFs.exists(fileToCopy) || needCopy(this.fs.getFileStatus(fileToCopy), targetFs.getFileStatus(fileToCopy))) {
```

Comment:
Without calling .exist to make sure file exist, if we directly call getFileStatus, it will throw FileNotFoundException when file not exist, I think we should avoid that situation? Unless we catch that exception and use it to determine whether file exist or not, which does not look like a correct approach for me.

Reply:
ok, I agree that creating an exception stacktrace is too costly.  (I thought it might return `null`.)  seems you'll need to call both.  I might abstract into a utility:
```
Optional<FileStatus> getOptFileStatus(FileSystem fs, Path p)
```
your choice